File: dotnet/test/Zametek.Maths.Graphs.Compilers.Tests/Engines/IncrementalCriticalPathTests.cs
using Shouldly;
using System;
using System.Collections.Generic;
using System.Linq;
using Xunit;

namespace Zametek.Maths.Graphs.Tests
{
    // The safety net for the incremental critical-path calculation.
    //
    // The priority-list calculation changes one activity's duration per iteration and
    // recalculates. Doing that incrementally is only worth anything if it is exactly
    // equivalent to recalculating in full, and "exactly" means every value, not just the
    // ones the priority-list selection happens to read - a future caller may read the
    // rest, and the two paths must not drift.
    //
    // So these compare the incremental path against the full one directly, rather than
    // against a recorded baseline: the full passes are the specification, and the test
    // runs both and demands they agree. The committed priority-list baseline covers the
    // end result on top of this.
    public class IncrementalCriticalPathTests
    {
        // Declines to begin an incremental calculation, so the builder falls back to
        // recalculating in full after every duration change - the behaviour these tests
        // compare against.
        private sealed class FullRecalculationOnlyEngine
            : IVertexCriticalPathEngine<int, int, int, IDependentActivity<int, int, int>>
        {
            private readonly IVertexCriticalPathEngine<int, int, int, IDependentActivity<int, int, int>> m_Inner =
                new VertexCriticalPathEngine<int, int, int, IDependentActivity<int, int, int>>();

            public bool CalculateCriticalPathForwardFlow(
                IVertexGraphState<int, int, int, IDependentActivity<int, int, int>> state,
                List<IInvalidConstraint<int>> invalidConstraints,
                bool shuffle) =>
                m_Inner.CalculateCriticalPathForwardFlow(state, invalidConstraints, shuffle);

            public bool CalculateCriticalPathBackwardFlow(
                IVertexGraphState<int, int, int, IDependentActivity<int, int, int>> state,
                List<IInvalidConstraint<int>> invalidConstraints,
                bool shuffle) =>
                m_Inner.CalculateCriticalPathBackwardFlow(state, invalidConstraints, shuffle);

            public bool BackFillIsolatedNodes(
                IVertexGraphState<int, int, int, IDependentActivity<int, int, int>> state,
                List<IInvalidConstraint<int>> invalidConstraints) =>
                m_Inner.BackFillIsolatedNodes(state, invalidConstraints);

            public IVertexIncrementalCriticalPath<int> BeginIncrementalCriticalPath(
                IVertexGraphState<int, int, int, IDependentActivity<int, int, int>> state) => null;
        }

        // Captures the graph state on its way through, so a test can drive an incremental
        // session against the same graph a builder is using without the builder needing to
        // expose its state.
        private sealed class StateCapturingEngine
            : IVertexCriticalPathEngine<int, int, int, IDependentActivity<int, int, int>>
        {
            private readonly IVertexCriticalPathEngine<int, int, int, IDependentActivity<int, int, int>> m_Inner =
                new VertexCriticalPathEngine<int, int, int, IDependentActivity<int, int, int>>();

            internal IVertexGraphState<int, int, int, IDependentActivity<int, int, int>> State { get; private set; }

            public bool CalculateCriticalPathForwardFlow(
                IVertexGraphState<int, int, int, IDependentActivity<int, int, int>> state,
                List<IInvalidConstraint<int>> invalidConstraints,
                bool shuffle)
            {
                State = state;
                return m_Inner.CalculateCriticalPathForwardFlow(state, invalidConstraints, shuffle);
            }

            public bool CalculateCriticalPathBackwardFlow(
                IVertexGraphState<int, int, int, IDependentActivity<int, int, int>> state,
                List<IInvalidConstraint<int>> invalidConstraints,
                bool shuffle) =>
                m_Inner.CalculateCriticalPathBackwardFlow(state, invalidConstraints, shuffle);

            public bool BackFillIsolatedNodes(
                IVertexGraphState<int, int, int, IDependentActivity<int, int, int>> state,
                List<IInvalidConstraint<int>> invalidConstraints) =>
                m_Inner.BackFillIsolatedNodes(state, invalidConstraints);

            public IVertexIncrementalCriticalPath<int> BeginIncrementalCriticalPath(
                IVertexGraphState<int, int, int, IDependentActivity<int, int, int>> state) =>
                m_Inner.BeginIncrementalCriticalPath(state);
        }

        private static VertexGraphBuilder<int, int, int, IDependentActivity<int, int, int>> Build(
            CorpusShapes.GraphSpec spec,
            IVertexCriticalPathEngine<int, int, int, IDependentActivity<int, int, int>> engine = null)
        {
            var graphBuilder = new VertexGraphBuilder<int, int, int, IDependentActivity<int, int, int>>(
                new VertexGraphBuilderEngines<int, int, int, IDependentActivity<int, int, int>>
                {
                    EdgeIdGenerator = new NextIdGenerator<int>(1_000_000),
                    CriticalPathEngine = engine
                        ?? new VertexCriticalPathEngine<int, int, int, IDependentActivity<int, int, int>>(),
                });

            foreach (CorpusShapes.ActivitySpec activity in spec.Activities)
            {
                var dependentActivity = new DependentActivity<int, int, int>(activity.Id, activity.Duration)
                {
                    MinimumEarliestStartTime = activity.MinimumEarliestStartTime,
                    MinimumFreeSlack = activity.MinimumFreeSlack,
                    MaximumLatestFinishTime = activity.MaximumLatestFinishTime,
                };

                graphBuilder.AddActivity(dependentActivity, new HashSet<int>(activity.Dependencies));
            }

            return graphBuilder;
        }

        [Fact]
        public void PriorityList_GivenCorpus_ThenIncrementalMatchesFullRecalculation()
        {
            foreach (CorpusShapes.GraphSpec spec in CorpusShapes.Generate())
            {
                List<int> incremental = Build(spec).CalculateCriticalPathPriorityList();
                List<int> full = Build(spec, new FullRecalculationOnlyEngine()).CalculateCriticalPathPriorityList();

                incremental.ShouldBe(full, $@"case {spec.Name}");
            }
        }

        [Theory]
        // Deliberately larger and deeper than the corpus, because the corpus graphs are
        // small enough that a propagation bug could stop before it reached anything.
        [InlineData(200, 8)]
        [InlineData(200, 40)]
        [InlineData(400, 20)]
        [InlineData(400, 100)]
        public void PriorityList_GivenLayeredGraph_ThenIncrementalMatchesFullRecalculation(int size, int layers)
        {
            CorpusShapes.GraphSpec spec = LayeredSpec(size, layers);

            List<int> incremental = Build(spec).CalculateCriticalPathPriorityList();
            List<int> full = Build(spec, new FullRecalculationOnlyEngine()).CalculateCriticalPathPriorityList();

            incremental.ShouldBe(full);
            incremental.Count.ShouldBe(size);
        }

        [Fact]
        public void CriticalPathValues_GivenCorpus_WhenDurationsAreZeroedOneAtATime_ThenIncrementalMatchesFullRecalculation()
        {
            // The strongest of these: it compares every value the two flows produce, not
            // just the priority list, after every single duration change. Free slack is
            // included, which the priority-list selection never reads.
            foreach (CorpusShapes.GraphSpec spec in CorpusShapes.Generate())
            {
                var capturingEngine = new StateCapturingEngine();
                VertexGraphBuilder<int, int, int, IDependentActivity<int, int, int>> incremental =
                    Build(spec, capturingEngine);
                VertexGraphBuilder<int, int, int, IDependentActivity<int, int, int>> full = Build(spec);

                incremental.CalculateCriticalPath();
                full.CalculateCriticalPath();

                IVertexIncrementalCriticalPath<int> session =
                    capturingEngine.BeginIncrementalCriticalPath(capturingEngine.State);
                session.ShouldNotBeNull($@"case {spec.Name}");

                foreach (int activityId in spec.Activities.Select(x => x.Id).OrderBy(x => x))
                {
                    incremental.Activity(activityId).Duration = 0;
                    full.Activity(activityId).Duration = 0;

                    if (!session.ApplyDurationChange(activityId))
                    {
                        incremental.CalculateCriticalPath();
                    }
                    full.CalculateCriticalPath();

                    AssertSameValues(incremental, full, $@"case {spec.Name}, after zeroing {activityId}");
                }
            }
        }

        private static void AssertSameValues(
            VertexGraphBuilder<int, int, int, IDependentActivity<int, int, int>> incremental,
            VertexGraphBuilder<int, int, int, IDependentActivity<int, int, int>> full,
            string because)
        {
            Render(incremental).ShouldBe(Render(full), because);
        }

        // Every value the two flows are responsible for, ordered by ID so the comparison
        // does not depend on iteration order.
        private static string Render(
            VertexGraphBuilder<int, int, int, IDependentActivity<int, int, int>> graphBuilder)
        {
            string activities = string.Join(
                @",",
                graphBuilder.Activities
                    .OrderBy(x => x.Id)
                    .Select(x => $@"{x.Id}={Show(x.EarliestStartTime)}/{Show(x.LatestFinishTime)}/{Show(x.FreeSlack)}"));

            string events = string.Join(
                @",",
                graphBuilder.Events
                    .OrderBy(x => x.Id)
                    .Select(x => $@"{x.Id}={Show(x.EarliestFinishTime)}/{Show(x.LatestFinishTime)}"));

            return $@"activities {activities} events {events}";
        }

        private static string Show(int? value) => value.HasValue ? value.Value.ToString() : @"-";

        // Each activity depends on two in the previous layer, which gives the propagation
        // somewhere to fan out to and somewhere to stop.
        private static CorpusShapes.GraphSpec LayeredSpec(int size, int layers)
        {
            var activities = new List<CorpusShapes.ActivitySpec>();
            int perLayer = Math.Max(1, size / layers);

            for (int id = 1; id <= size; id++)
            {
                int layer = (id - 1) / perLayer;
                var dependencies = new HashSet<int>();

                if (layer > 0)
                {
                    int previousStart = ((layer - 1) * perLayer) + 1;
                    int previousEnd = Math.Min(layer * perLayer, size);

                    if (previousEnd >= previousStart)
                    {
                        dependencies.Add(previousStart + ((id * 7) % (previousEnd - previousStart + 1)));
                        dependencies.Add(previousStart + ((id * 13) % (previousEnd - previousStart + 1)));
                    }
                }

                var spec = new CorpusShapes.ActivitySpec(id, 1 + (id % 9), dependencies);

                // A scattering of constraints, because they are what make the two passes
                // interesting - the maximum latest finish time in particular feeds the
                // earliest start time through the duration, so zeroing an activity can
                // raise its own earliest start rather than only lowering its finish.
                if (id % 17 == 0)
                {
                    spec.MinimumEarliestStartTime = id % 5;
                }
                if (id % 23 == 0)
                {
                    spec.MaximumLatestFinishTime = 400 + (id % 11);
                }
                else if (id % 29 == 0)
                {
                    spec.MinimumFreeSlack = id % 4;
                }

                activities.Add(spec);
            }

            return new CorpusShapes.GraphSpec($@"layered-{size}-{layers}", activities);
        }
    }
}

File: rust/compilers/tests/incremental_critical_path_tests.rs
//! Port of `IncrementalCriticalPathTests.cs`.
//!
//! The priority-list calculation changes one activity's duration per iteration and
//! recalculates. Doing that incrementally is only worth anything if it is exactly
//! equivalent to recalculating in full, so these compare the two directly rather
//! than against a recorded baseline: the full passes are the specification, and the
//! test runs both and demands they agree.
//!
//! The C# suite additionally compares the two paths value for value after every
//! single duration change, including free slack, which the priority-list selection
//! never reads. That check needs to drive a session against a builder's own graph
//! state, which is private here, so it is not duplicated - and it does not need to
//! be: both languages check their priority lists against the same committed
//! baseline, so a divergence in the shared algorithm would fail there.

mod corpus_shapes;

use indexmap::IndexSet;
use std::sync::Arc;
use zametek_maths_graphs_compilers::contracts::IVertexCriticalPathEngine;
use zametek_maths_graphs_compilers::vertex::{IncrementalCriticalPath, VertexGraphState};
use zametek_maths_graphs_compilers::{
    NextIdGenerator, VertexCriticalPathEngine, VertexGraphBuilder, VertexGraphBuilderEngines,
};
use zametek_maths_graphs_primitives::{DependentActivity, GraphError, InvalidConstraint};

/// Declines to begin an incremental calculation, so the builder falls back to
/// recalculating in full after every duration change - the behaviour these tests
/// compare against.
#[derive(Default)]
struct FullRecalculationOnlyEngine {
    inner: VertexCriticalPathEngine,
}

impl IVertexCriticalPathEngine<i32, i32, i32> for FullRecalculationOnlyEngine {
    fn calculate_critical_path_forward_flow(
        &self,
        state: &mut VertexGraphState<i32, i32, i32>,
        invalid_constraints: &[InvalidConstraint<i32>],
        shuffle: bool,
    ) -> Result<bool, GraphError> {
        self.inner
            .calculate_critical_path_forward_flow(state, invalid_constraints, shuffle)
    }

    fn calculate_critical_path_backward_flow(
        &self,
        state: &mut VertexGraphState<i32, i32, i32>,
        invalid_constraints: &[InvalidConstraint<i32>],
        shuffle: bool,
    ) -> Result<bool, GraphError> {
        self.inner
            .calculate_critical_path_backward_flow(state, invalid_constraints, shuffle)
    }

    fn back_fill_isolated_nodes(
        &self,
        state: &mut VertexGraphState<i32, i32, i32>,
        invalid_constraints: &[InvalidConstraint<i32>],
    ) -> bool {
        self.inner
            .back_fill_isolated_nodes(state, invalid_constraints)
    }

    fn begin_incremental_critical_path(
        &self,
        _state: &VertexGraphState<i32, i32, i32>,
    ) -> Option<IncrementalCriticalPath<i32>> {
        None
    }
}

fn build(spec: &corpus_shapes::GraphSpec, full_only: bool) -> VertexGraphBuilder<i32, i32, i32> {
    let mut engines = VertexGraphBuilderEngines::<i32, i32, i32> {
        edge_id_generator: Box::new(NextIdGenerator::new(1_000_000)),
        ..Default::default()
    };

    if full_only {
        engines.critical_path_engine = Arc::new(FullRecalculationOnlyEngine::default());
    }

    let mut graph_builder = VertexGraphBuilder::with_engines(engines);

    for activity_spec in &spec.activities {
        let mut activity = DependentActivity::new(activity_spec.id, activity_spec.duration);
        activity.minimum_earliest_start_time = activity_spec.minimum_earliest_start_time;
        activity.minimum_free_slack = activity_spec.minimum_free_slack;
        activity.maximum_latest_finish_time = activity_spec.maximum_latest_finish_time;
        graph_builder.add_activity_with_dependencies(activity, activity_spec.dependencies.clone());
    }

    graph_builder
}

#[test]
fn priority_list_given_corpus_then_incremental_matches_full_recalculation() {
    for spec in corpus_shapes::generate() {
        let incremental = build(&spec, false)
            .calculate_critical_path_priority_list()
            .unwrap_or_else(|error| panic!("case {} failed: {error}", spec.name));
        let full = build(&spec, true)
            .calculate_critical_path_priority_list()
            .unwrap_or_else(|error| panic!("case {} failed: {error}", spec.name));

        assert_eq!(incremental, full, "case {}", spec.name);
    }
}

#[test]
fn priority_list_given_layered_graphs_then_incremental_matches_full_recalculation() {
    // Deliberately larger and deeper than the corpus, because the corpus graphs are
    // small enough that a propagation bug could stop before it reached anything.
    for (size, layers) in [(200, 8), (200, 40), (400, 20), (400, 100)] {
        let spec = layered_spec(size, layers);

        let incremental = build(&spec, false)
            .calculate_critical_path_priority_list()
            .expect("incremental priority list");
        let full = build(&spec, true)
            .calculate_critical_path_priority_list()
            .expect("full priority list");

        assert_eq!(incremental, full, "layered {size}/{layers}");
        assert_eq!(incremental.len(), size as usize, "layered {size}/{layers}");
    }
}

/// Each activity depends on two in the previous layer, which gives the propagation
/// somewhere to fan out to and somewhere to stop.
fn layered_spec(size: i32, layers: i32) -> corpus_shapes::GraphSpec {
    let mut activities: Vec<corpus_shapes::ActivitySpec> = Vec::new();
    let per_layer = (size / layers).max(1);

    for id in 1..=size {
        let layer = (id - 1) / per_layer;
        let mut dependencies: IndexSet<i32> = IndexSet::new();

        if layer > 0 {
            let previous_start = ((layer - 1) * per_layer) + 1;
            let previous_end = (layer * per_layer).min(size);

            if previous_end >= previous_start {
                let span = previous_end - previous_start + 1;
                dependencies.insert(previous_start + ((id * 7) % span));
                dependencies.insert(previous_start + ((id * 13) % span));
            }
        }

        let mut spec = corpus_shapes::ActivitySpec {
            id,
            duration: 1 + (id % 9),
            dependencies,
            minimum_earliest_start_time: None,
            minimum_free_slack: None,
            maximum_latest_finish_time: None,
        };

        // A scattering of constraints, because they are what make the two passes
        // interesting - the maximum latest finish time in particular feeds the
        // earliest start time through the duration, so zeroing an activity can raise
        // its own earliest start rather than only lowering its finish.
        if id % 17 == 0 {
            spec.minimum_earliest_start_time = Some(id % 5);
        }
        if id % 23 == 0 {
            spec.maximum_latest_finish_time = Some(400 + (id % 11));
        } else if id % 29 == 0 {
            spec.minimum_free_slack = Some(id % 4);
        }

        activities.push(spec);
    }

    corpus_shapes::GraphSpec {
        name: format!("layered-{size}-{layers}"),
        activities,
    }
}

#[test]
fn priority_list_given_chain_then_incremental_matches_full_recalculation() {
    // A pure chain is the shape where every selected activity is on the critical path,
    // so the project finish time moves on nearly every iteration - the case the
    // propagation seeds every End node for rather than giving up on.
    let mut activities: Vec<corpus_shapes::ActivitySpec> = Vec::new();

    for id in 1..=150 {
        let mut dependencies: IndexSet<i32> = IndexSet::new();
        if id > 1 {
            dependencies.insert(id - 1);
        }
        activities.push(corpus_shapes::ActivitySpec {
            id,
            duration: 1 + (id % 7),
            dependencies,
            minimum_earliest_start_time: None,
            minimum_free_slack: None,
            maximum_latest_finish_time: None,
        });
    }

    let spec = corpus_shapes::GraphSpec {
        name: "chain-150".to_string(),
        activities,
    };

    let incremental = build(&spec, false)
        .calculate_critical_path_priority_list()
        .expect("incremental priority list");
    let full = build(&spec, true)
        .calculate_critical_path_priority_list()
        .expect("full priority list");

    assert_eq!(incremental, full);
}

#[test]
fn priority_list_given_isolated_activities_then_incremental_matches_full_recalculation() {
    // Isolated nodes are the ones the backward pass never reaches, so their latest
    // finish time is maintained by the forward propagation alone.
    let mut activities: Vec<corpus_shapes::ActivitySpec> = Vec::new();

    for id in 1..=20 {
        activities.push(corpus_shapes::ActivitySpec {
            id,
            duration: 1 + (id % 5),
            dependencies: IndexSet::new(),
            minimum_earliest_start_time: None,
            minimum_free_slack: None,
            maximum_latest_finish_time: if id % 4 == 0 { Some(10) } else { None },
        });
    }

    let spec = corpus_shapes::GraphSpec {
        name: "isolated-20".to_string(),
        activities,
    };

    let incremental = build(&spec, false)
        .calculate_critical_path_priority_list()
        .expect("incremental priority list");
    let full = build(&spec, true)
        .calculate_critical_path_priority_list()
        .expect("full priority list");

    assert_eq!(incremental, full);
}
